From 2f00b83720aa796f14c7dea237b28fcb90699e49 Mon Sep 17 00:00:00 2001
From: CL Wang <cl634@andestech.com>
Date: Wed, 17 Apr 2024 14:54:57 +0800
Subject: [PATCH 048/168] dmaengine: andes: atcdmac300g: Added new variable
 "used_chan" to determine the channels used by the DMA engine driver

This patch contains the item 2 of the following modifications

On the AE350 platform, the ftssp does not have a dedicated DMA hardware
and requires the DAM engine to transfer data. Before this modification,
the architecture used was the FTSSP driver combined with the legacy DMA
driver to transfer audio data. However, the limitation of this method is
that the legacy DMA driver and the DMA engine driver cannot operate
simultaneously. If the FTSSP were to use the DMA engine driver for data
transfer, it would require a significant amount of development time.
Therefore, we opted to use this patch to minimize modifications and reduce
the required development time.

Objectives of this Modification
 - Rename the legacy DMA driver to ftssp_dma and place it in the same
   directory as FTSSP, making it exclusive to FTSSP.
 - The DMA engine driver can be compiled and assembled independently.
 - Allow the DMA engine driver to operate independently, regardless of
   the presence of ftssp_dma.
 - Ensure that when ftssp_dma is operating, it does not affect the current
   operation of the DMA engine driver.

Operation of the ftssp_dma and the atcdmac300g DMA Engine Driver
1. DMA Hardware Operation:
   Both the atcdmac300g DMA engine driver and ftssp_dma driver co-exist in
   the system and share the DMA hardware.

   a. The atcdmac300g DMA engine driver registers with the DMA framework
      and provides all DMA engine operations.

   b. ftssp_dma uses the DMA engine API (dma_request_channel) to obtain an
      available DMA channel from the DMA engine driver. It then configures
      the hardware resources of this channel. ftssp_dma controls only the
      hardware resources of the allocated channel, ensuring no hardware
      conflicts as each manages its allocated DMA channels.

2. IRQ Handling:
   After data transfer is completed, the DMA hardware triggers an
   interrupt. The following modifications ensure that the interrupt is
   handled by the appropriate ISR. Interrupts triggered by the DMA engine
   driver are handled by the DMA engine driver, while interrupts for FTSSP
   data transfer are handled by ftssp_dma.

   a. How to determine if the interrupt belongs to the DMA engine driver
      or FTSSP driver?
      A new variable used_chan is added to the struct v5_dma in the DMA
      engine driver to determine which channels the DMA engine driver
      uses. used_chan records the channels currently being used by the
      DMA engine driver for data transfer. Each bit in used_chan
      represents a channel; a value of 1 indicates that the channel is
      being used by the DMA engine driver. In the ISR, the values of
      used_chan and the interrupt status are used to determine which
      channels are being used for data transfer and which have completed
      the transfer in the DMA engine driver. Interrupts not belonging to
      the DMA engine driver are handed over to ftssp_dma.
   b. How to hand over interrupts to ftssp_dma?
      Using the shared IRQ mechanism provided by Linux, multiple ISRs can
      be registered. Both the DMA engine and FTSSP will register an ISR
      with the IRQF_SHARED attribute. Using the mechanism described in
      part (a), each ISR handles its triggered interrupts independently.

3. Changing Legacy DMA to ftssp_dma:
   In order to make the legacy DMA driver exclusive to FTSSP, the following
   changes were made

   a. Remove all EXPORT_SYMBOL APIs from the legacy DMA driver.
   b. Remove the registration of the platform driver and platform device.
   c. Since the ftssp_dma driver needs to use functions provided by the
      DMA engine driver, its dependency is changed to ATCDMAC300G.

Signed-off-by: CL Wang <cl634@andestech.com>
---
 drivers/dma/atcdmac300g.c | 51 +++++++++++++++++++++++++++------------
 drivers/dma/atcdmac300g.h |  5 ++++
 2 files changed, 40 insertions(+), 16 deletions(-)

diff --git a/drivers/dma/atcdmac300g.c b/drivers/dma/atcdmac300g.c
index 5c92d75bf4d8..614010da19f0 100644
--- a/drivers/dma/atcdmac300g.c
+++ b/drivers/dma/atcdmac300g.c
@@ -201,6 +201,9 @@ static void v5_desc_chain(struct v5_desc **first, struct v5_desc **prev,
  */
 static void v5_dostart(struct v5_dma_chan *v5chan, struct v5_desc *first)
 {
+	struct v5_dma *v5dma = v5chan->device;
+	unsigned long dma_flags;
+
 	if (v5_chan_is_enabled(v5chan)) {
 		dev_err(chan2dev(&v5chan->chan_common),
 			"BUG: Attempted to start non-idle channel\n");
@@ -213,6 +216,10 @@ static void v5_dostart(struct v5_dma_chan *v5chan, struct v5_desc *first)
 			v5_channel_readl(v5chan, CH_LLP_LOW_OFF));
 		return;
 	}
+	spin_lock_irqsave(&v5dma->dma_lock, dma_flags);
+	v5dma->used_chan |= (1 << v5chan->chan_id);
+	spin_unlock_irqrestore(&v5dma->dma_lock, dma_flags);
+
 	vdbg_dump_regs(v5chan);
 	v5_channel_writel(v5chan, CH_CTL_OFF, first->lli.ctrl);
 	v5_channel_writel(v5chan, CH_SIZE_OFF, first->lli.tranSize);
@@ -391,34 +398,45 @@ static irqreturn_t v5_dma_interrupt(int irq, void *dev_id)
 {
 	struct v5_dma		*v5dma = (struct v5_dma *)dev_id;
 	struct v5_dma_chan	*v5chan;
+	unsigned long		dma_flags;
 	int			i;
 	u32			status;
 	int			ret = IRQ_NONE;
+	u32			int_ch;
+	u16			used_ch;
 
-	do {
-		status = v5_dma_readl(v5dma, INT_STA);
-
-		dev_vdbg(v5dma->dma_common.dev,
-			"int: sta = 0x%08x\n", status);
-		if (status == 0)
-			break;
+	status = v5_dma_readl(v5dma, INT_STA);
+	used_ch = v5dma->used_chan;
+	int_ch = used_ch & V5_DMA_INT_ALL(status);
 
-		v5_dma_writel(v5dma, INT_STA, status);
+	while (int_ch) {
+		spin_lock_irqsave(&v5dma->dma_lock, dma_flags);
+		v5dma->used_chan &= ~int_ch;
+		spin_unlock_irqrestore(&v5dma->dma_lock, dma_flags);
+		v5_dma_writel(v5dma, INT_STA, V5_DMA_INT_CLR(int_ch));
 
 		for (i = 0; i < v5dma->dma_common.chancnt; i++) {
-			v5chan = &v5dma->chan[i];
-			if (status & (V5_DMA_TC(i)))
-				set_bit(V5_IS_TC, &v5chan->status);
+			if (int_ch & (1 << i)) {
+				int_ch &= ~(1 << i);
+				v5chan = &v5dma->chan[i];
+
+				if (status & (V5_DMA_TC(i)))
+					set_bit(V5_IS_TC, &v5chan->status);
 
-			if (status & (V5_DMA_ABT(i) | V5_DMA_ERR(i)))
-				set_bit(V5_IS_ERR, &v5chan->status);
+				if (status & (V5_DMA_ABT(i) | V5_DMA_ERR(i)))
+					set_bit(V5_IS_ERR, &v5chan->status);
 
-			if (v5chan->status) {
 				tasklet_schedule(&v5chan->tasklet);
 				ret = IRQ_HANDLED;
 			}
+			if (!int_ch)
+				break;
 		}
-	} while (status);
+
+		status = v5_dma_readl(v5dma, INT_STA);
+		used_ch = v5dma->used_chan;
+		int_ch = used_ch & V5_DMA_INT_ALL(status);
+	};
 
 	return ret;
 }
@@ -1143,6 +1161,7 @@ static int __init v5_dma_probe(struct platform_device *pdev)
 	if (!v5dma)
 		return -ENOMEM;
 
+	spin_lock_init(&v5dma->dma_lock);
 	v5dma->io_regs = 0;
 	if (dev_is_dma_coherent(&pdev->dev)) {
 		u64 taddr;
@@ -1174,7 +1193,7 @@ static int __init v5_dma_probe(struct platform_device *pdev)
 		v5dma->ch : plat_dat->nr_channels;
 
 	v5_dma_off(v5dma);
-	err = request_irq(irq, v5_dma_interrupt, 0, "v5_dmac", v5dma);
+	err = request_irq(irq, v5_dma_interrupt, IRQF_SHARED, "v5_dmac", v5dma);
 	if (err)
 		goto err_irq;
 
diff --git a/drivers/dma/atcdmac300g.h b/drivers/dma/atcdmac300g.h
index fdb70903cd2f..dcc4a6446f37 100644
--- a/drivers/dma/atcdmac300g.h
+++ b/drivers/dma/atcdmac300g.h
@@ -41,6 +41,9 @@ static inline void clrbl(addr_t bit, void __iomem *reg)
 #define	V5_DMA_TC(x)			(0x1 << (TC_OFFSET + (x)))
 #define	V5_DMA_ABT(x)			(0x1 << (ABT_OFFSET + (x)))
 #define	V5_DMA_ERR(x)			(0x1 << (ERR_OFFSET + (x)))
+#define	V5_DMA_INT_ALL(x)		(((x >> TC_OFFSET) & 0xFF) | ((x >> ABT_OFFSET) & 0xFF) \
+					 | ((x >> ERR_OFFSET) & 0xFF))
+#define	V5_DMA_INT_CLR(x)		((x << TC_OFFSET) | (x << ABT_OFFSET) | (x << ERR_OFFSET))
 /* Channel Enable Register */
 #define CH_EN				(DMAC_BASE + 0x34)
 /* Channel Base Register */
@@ -303,7 +306,9 @@ struct v5_dma {
 	struct dma_device	dma_common;
 	void __iomem		*regs;
 	void __iomem		*io_regs;
+	spinlock_t		dma_lock;
 	u32			ctl;
+	u16			used_chan;
 	u8			data_width;
 	u8			ch;
 	u8			all_chan_mask;
-- 
2.34.1

